receiver/awscontainerinsightreceiver/internal/k8sapiserver/k8sapiserver.go

```diff
@@ -92,10 +89,16 @@ type clusterNameProvider interface {
 	GetClusterName() string
 }
 
-type k8sAPIServerOption func(*K8sAPIServer)
+type Option func(*K8sAPIServer)
+
+func WithLeaderLockName(name string) Option {
+	return func(server *K8sAPIServer) {
+		server.leaderLockName = name
+	}
+}
 
 // New creates a k8sApiServer which can generate cluster-level metrics
-func New(clusterNameProvider clusterNameProvider, logger *zap.Logger, options ...k8sAPIServerOption) (*K8sAPIServer, error) {
+func New(clusterNameProvider clusterNameProvider, logger *zap.Logger, options ...Option) (*K8sAPIServer, error) {
```

Comment:
I think the option pattern is good here, but it looks like `leaderLockName` is a required during `init`, so we should set a default value in the struct creation to ensure it can't be empty string.  If empty string is an appropriate default value then nevermind.

Reply:
Hi @TylerHelmuth , empty string can cause issues. However, just wanted to clarify a bit more.

Please correct me if Im wrong but I think [init()](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21034/files/49b2e4c2d704a88706b3d73a90237a4909d8861f#diff-1a583ee28f5c5b578fb9623462476ff5b07e8f862be6465571fbf4a7c731e9d1R207) in this case is associated with `K8sAPIServer` and hence wont get invoked during package import but has to be explicitly called.

I see it being invoked [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/49b2e4c2d704a88706b3d73a90237a4909d8861f/receiver/awscontainerinsightreceiver/internal/k8sapiserver/k8sapiserver.go#L117), and by this point, the config should have been initialized by the factory based on whatever is specified either by the user or by the [default](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21034/files/49b2e4c2d704a88706b3d73a90237a4909d8861f#diff-20e55430a41b074dab0e6cc3ef2551146c6856cd76d2d3aff3f2450099acb0f4R50). So ideally there should never be a case where it is empty.

Reply:
You are correct that all the current uses of `k8sapiserver.New` use the `k8sapiserver.WithLeaderLockName` option with a guaranteed value because of the default config function so the component is indeed safe.  But from an API perspective, if somewhere were to use `k8sapiserver.New` somewhere else in the component they could put themselves in a dangerous situation.

It is definitely a nit, but since the internal function could be used anywhere in the component we should ensure it doesn't let people hurt themselves.  At the moment `New` must be called with `k8sapiserver.WithLeaderLockName`, which makes that less of an option and more of a requirement.

Reply:
Ah I now see what you mean! That definitely makes sense. I guess its best to safeguard against that so probably defining a default [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21034/files#diff-1a583ee28f5c5b578fb9623462476ff5b07e8f862be6465571fbf4a7c731e9d1R75) would be best. I can update that.